# File: 08_methylation_extractor_array_job.R
# Auth: umar.niazi@kcl.as.uk
# DESC: create a parameter file and shell script to run array job on hpc
# Date: 23/08/2017


## set variables and source libraries
source('header.R')

## connect to mysql database to get sample information
library('RMySQL')

##### connect to mysql database to get samples
db = dbConnect(MySQL(), user='rstudio', password='12345', dbname='Projects', host='127.0.0.1')
dbListTables(db)

# check how many files each sample has
g_did
q = paste0('select Sample.id as sid, Sample.group1, Sample.group2, Sample.group3, Sample.title, File.* from Sample, File
           where (Sample.idData = 14) AND (File.idSample = Sample.id AND File.type like "%duplicates removed%")')
dfSample = dbGetQuery(db, q)
nrow(dfSample)
dfSample
# close connection after getting data
dbDisconnect(db)
# remove any whitespace from the names
dfSample$name = gsub(" ", "", dfSample$name, fixed = T)
dfSample$title = gsub(" ", "", dfSample$title, fixed = T)

# set header variables 
cvShell = '#!/bin/bash'
cvShell.2 = '#$ -S /bin/bash'
cvProcessors = '#$ -pe smp 8'
cvWorkingDir = '#$ -cwd'
cvJobName = '#$ -N meth-ex-array'
cvStdout = '#$ -j y'
cvMemoryReserve = '#$ -l h_vmem=19G'
cvArrayJob = paste0('#$ -t 1-', nrow(dfSample))
# using high memory queue with one slot and 19 Gigs of memory

# set the directory names
cvInput = 'input/'
cvOutput = 'output/methylationExtractor/'
cvBismark = '/users/k1625253/brc_scratch/Programs/bismark_v0.18.1/'
cvGeneIndex = '/users/k1625253/brc_scratch/Data/MetaData/GenomeIndex/mm10_bismark/'

# create a parameter file and shell script
dir.create('AutoScripts')
oFile.param = file('AutoScripts/methylation_extractor_param.txt', 'wt')

temp = sapply(1:nrow(dfSample), function(x){
  # get the file names
  csFile = dfSample$name[x] 
  # remove white space from title
  csFile = gsub(" ", "", csFile, fixed=T)
  # write parameter file, the inputs
  in.r1 = paste0(cvInput, csFile)
  p1 = paste(in.r1)
  writeLines(p1, oFile.param)
})

close(oFile.param)

oFile = file('AutoScripts/methylation_extractor.sh', 'wt')

writeLines(c('# Autogenerated script from 08_methylation_extractor_array_job.R', paste('# date', date())), oFile)
writeLines(c('# make sure directory paths exist before running script'), oFile)
writeLines(c(cvShell, cvShell.2, cvProcessors, cvWorkingDir, cvJobName, cvStdout, cvMemoryReserve, cvArrayJob), oFile)
writeLines('\n\n', oFile)

# module load
writeLines(c('module load bioinformatics/samtools/1.3.1'), oFile)
writeLines('\n\n', oFile)

## write array job lines
writeLines("# Parse parameter file to get variables.
number=$SGE_TASK_ID
paramfile=methylation_extractor_param.txt
 
inr1=`sed -n ${number}p $paramfile | awk '{print $1}'`

# 9. Run the program.", oFile)
cvProg = paste(cvBismark, 'bismark_methylation_extractor', sep='')
p1 = paste(cvProg, '-p --gzip --parallel 8 --report -o', cvOutput, '$inr1', sep=' ')
com = paste(p1)

writeLines(com, oFile)
writeLines('\n\n', oFile)

close(oFile)
#dbDisconnect(db)
